File: src/claw_mem/multimodal/multimodal_memory.py
"""
Multimodal Memory Module for claw-mem v2.4.0

Supports multimodal memory storage for images, files, etc.
"""

import hashlib
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from pathlib import Path
from typing import Any, Dict, List, Optional


class MemoryType(Enum):
    """Memory type"""

    TEXT = "text"
    IMAGE = "image"
    FILE = "file"
    AUDIO = "audio"
    VIDEO = "video"


@dataclass
class ImageMemory:
    """Image memory"""

    image_id: str
    description: str
    path: str
    thumbnail_path: Optional[str] = None
    metadata: Dict[str, Any] = field(default_factory=dict)
    created_at: datetime = field(default_factory=datetime.now)


@dataclass
class FileMemory:
    """File memory"""

    file_id: str
    filename: str
    file_type: str
    path: str
    size_bytes: int = 0
    metadata: Dict[str, Any] = field(default_factory=dict)
    created_at: datetime = field(default_factory=datetime.now)


class MultimodalMemoryStore:
    """
    Multimodal memory storage

    Supports memory storage for non-text content such as images and files.
    """

    def __init__(self, base_path: str = "./workspace/multimodal"):
        self.base_path = Path(base_path)
        self.base_path.mkdir(parents=True, exist_ok=True)

        # Storage for different types
        self._image_memories: Dict[str, ImageMemory] = {}
        self._file_memories: Dict[str, FileMemory] = {}

    def store_image(
        self, image_path: str, description: str, metadata: Optional[Dict[str, Any]] = None
    ) -> str:
        """
        Store image memory

        Args:
            image_path: Image file path
            description: Image description
            metadata: Additional metadata

        Returns:
            Image memory ID
        """
        # Generate ID from content hash
        image_id = self._generate_id(image_path)

        # Copy image to storage
        storage_path = self.base_path / "images" / Path(image_path).name
        storage_path.parent.mkdir(parents=True, exist_ok=True)

        # Note: In real implementation, would copy file here
        # For now, just store reference

        memory = ImageMemory(
            image_id=image_id,
            description=description,
            path=str(storage_path),
            metadata=metadata or {},
        )

        self._image_memories[image_id] = memory
        return image_id

    def store_file(
        self,
        file_path: str,
        file_type: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> str:
        """
        Store file memory

        Args:
            file_path: File path
            file_type: File type
            metadata: Additional metadata

        Returns:
            File memory ID
        """
        file_path_obj = Path(file_path)
        file_id = self._generate_id(file_path)

        # Determine file type
        if file_type is None:
            file_type = file_path_obj.suffix.lstrip(".")

        # Get file size
        size_bytes = 0
        if file_path_obj.exists():
            size_bytes = file_path_obj.stat().st_size

        # Copy to storage
        storage_path = self.base_path / "files" / file_path_obj.name
        storage_path.parent.mkdir(parents=True, exist_ok=True)

        memory = FileMemory(
            file_id=file_id,
            filename=file_path_obj.name,
            file_type=file_type,
            path=str(storage_path),
            size_bytes=size_bytes,
            metadata=metadata or {},
        )

        self._file_memories[file_id] = memory
        return file_id

    def get_image(self, image_id: str) -> Optional[ImageMemory]:
        """Get image memory"""
        return self._image_memories.get(image_id)

    def get_file(self, file_id: str) -> Optional[FileMemory]:
        """Get file memory"""
        return self._file_memories.get(file_id)

    def search_by_description(self, query: str, limit: int = 10) -> List[ImageMemory]:
        """Search images by description"""
        query_lower = query.lower()
        results = []

        for memory in self._image_memories.values():
            if query_lower in memory.description.lower():
                results.append(memory)

        return results[:limit]

    def get_stats(self) -> Dict[str, Any]:
        """Get statistics"""
        total_size = sum(f.size_bytes for f in self._file_memories.values())

        return {
            "total_images": len(self._image_memories),
            "total_files": len(self._file_memories),
            "total_size_bytes": total_size,
            "image_descriptions": [m.description for m in self._image_memories.values()],
        }

    def _generate_id(self, content: str) -> str:
        """Generate unique ID"""
        return hashlib.sha256(content.encode()).hexdigest()[:16]


# Global instance
_multimodal_store: Optional[MultimodalMemoryStore] = None


def get_multimodal_store(base_path: str = "./workspace/multimodal") -> MultimodalMemoryStore:
    """Get multimodal storage instance"""
    global _multimodal_store
    if _multimodal_store is None:
        _multimodal_store = MultimodalMemoryStore(base_path)
    return _multimodal_store


def reset_multimodal_store() -> None:
    """Reset global multimodal store instance (for testing)."""
    global _multimodal_store
    _multimodal_store = None

File: src/multimodal/multimodal_memory.ts
// Copyright 2026 Peter Cheng
// Licensed under Apache-2.0

/**
 * Multimodal Memory Module for claw-mem
 *
 * Supports multimodal memory storage for images, files, etc.
 */

import * as crypto from "crypto";
import * as fs from "fs";
import * as path from "path";

export enum MemoryType {
  TEXT = "text",
  IMAGE = "image",
  FILE = "file",
  AUDIO = "audio",
  VIDEO = "video",
}

export interface ImageMemoryData {
  image_id: string;
  description: string;
  path: string;
  thumbnail_path: string | null;
  metadata: Record<string, unknown>;
  created_at: string;
}

export class ImageMemory {
  imageId: string;
  description: string;
  path: string;
  thumbnailPath: string | null;
  metadata: Record<string, unknown>;
  createdAt: Date;

  constructor(
    imageId: string,
    description: string,
    path: string,
    thumbnailPath?: string | null,
    metadata?: Record<string, unknown>,
    createdAt?: Date,
  ) {
    this.imageId = imageId;
    this.description = description;
    this.path = path;
    this.thumbnailPath = thumbnailPath ?? null;
    this.metadata = metadata ?? {};
    this.createdAt = createdAt ?? new Date();
  }

  toDict(): ImageMemoryData {
    return {
      image_id: this.imageId,
      description: this.description,
      path: this.path,
      thumbnail_path: this.thumbnailPath,
      metadata: { ...this.metadata },
      created_at: this.createdAt.toISOString(),
    };
  }
}

export interface FileMemoryData {
  file_id: string;
  filename: string;
  file_type: string;
  path: string;
  size_bytes: number;
  metadata: Record<string, unknown>;
  created_at: string;
}

export class FileMemory {
  fileId: string;
  filename: string;
  fileType: string;
  path: string;
  sizeBytes: number;
  metadata: Record<string, unknown>;
  createdAt: Date;

  constructor(
    fileId: string,
    filename: string,
    fileType: string,
    path: string,
    sizeBytes: number = 0,
    metadata?: Record<string, unknown>,
    createdAt?: Date,
  ) {
    this.fileId = fileId;
    this.filename = filename;
    this.fileType = fileType;
    this.path = path;
    this.sizeBytes = sizeBytes;
    this.metadata = metadata ?? {};
    this.createdAt = createdAt ?? new Date();
  }

  toDict(): FileMemoryData {
    return {
      file_id: this.fileId,
      filename: this.filename,
      file_type: this.fileType,
      path: this.path,
      size_bytes: this.sizeBytes,
      metadata: { ...this.metadata },
      created_at: this.createdAt.toISOString(),
    };
  }
}

export interface MultimodalStats {
  total_images: number;
  total_files: number;
  total_size_bytes: number;
  image_descriptions: string[];
}

/**
 * Multimodal memory storage.
 *
 * Supports memory storage for non-text content such as images and files.
 */
export class MultimodalMemoryStore {
  basePath: string;

  private _imageMemories: Map<string, ImageMemory>;
  private _fileMemories: Map<string, FileMemory>;

  /**
   * @param basePath - Base directory for multimodal storage (default "./workspace/multimodal")
   */
  constructor(basePath: string = "./workspace/multimodal") {
    this.basePath = basePath;
    this._imageMemories = new Map();
    this._fileMemories = new Map();

    fs.mkdirSync(this.basePath, { recursive: true });
  }

  /**
   * Store an image memory.
   *
   * @param imagePath   - Image file path
   * @param description - Image description
   * @param metadata    - Additional metadata
   * @returns Image memory ID
   */
  storeImage(
    imagePath: string,
    description: string,
    metadata?: Record<string, unknown>,
  ): string {
    const imageId = this._generateId(imagePath);

    const storageDir = path.join(this.basePath, "images");
    const storagePath = path.join(storageDir, path.basename(imagePath));
    fs.mkdirSync(storageDir, { recursive: true });

    // Note: In real implementation, would copy file here.
    // For now, just store reference.

    const memory = new ImageMemory(imageId, description, storagePath, null, metadata ?? {});

    this._imageMemories.set(imageId, memory);
    return imageId;
  }

  /**
   * Store a file memory.
   *
   * @param filePath  - File path
   * @param fileType  - File type (inferred from extension if not provided)
   * @param metadata  - Additional metadata
   * @returns File memory ID
   */
  storeFile(
    filePath: string,
    fileType?: string,
    metadata?: Record<string, unknown>,
  ): string {
    const filePathObj = path.parse(filePath);
    const fileId = this._generateId(filePath);

    // Determine file type
    const resolvedFileType = fileType ?? filePathObj.ext.slice(1);

    // Get file size
    let sizeBytes = 0;
    try {
      const stat = fs.statSync(filePath);
      sizeBytes = stat.size;
    } catch {
      // file may not exist yet
    }

    // Copy to storage
    const storageDir = path.join(this.basePath, "files");
    const storagePath = path.join(storageDir, filePathObj.base);
    fs.mkdirSync(storageDir, { recursive: true });

    const memory = new FileMemory(
      fileId,
      filePathObj.base,
      resolvedFileType,
      storagePath,
      sizeBytes,
      metadata ?? {},
    );

    this._fileMemories.set(fileId, memory);
    return fileId;
  }

  /**
   * Get an image memory by ID.
   *
   * @param imageId - Image memory ID
   * @returns ImageMemory or undefined
   */
  getImage(imageId: string): ImageMemory | undefined {
    return this._imageMemories.get(imageId);
  }

  /**
   * Get a file memory by ID.
   *
   * @param fileId - File memory ID
   * @returns FileMemory or undefined
   */
  getFile(fileId: string): FileMemory | undefined {
    return this._fileMemories.get(fileId);
  }

  /**
   * Search images by description text.
   *
   * @param query - Search query
   * @param limit - Maximum results (default 10)
   * @returns Matching ImageMemory instances
   */
  searchByDescription(query: string, limit: number = 10): ImageMemory[] {
    const lower = query.toLowerCase();
    const results: ImageMemory[] = [];

    for (const memory of this._imageMemories.values()) {
      if (memory.description.toLowerCase().includes(lower)) {
        results.push(memory);
      }
    }

    return results.slice(0, limit);
  }

  /**
   * Get storage statistics.
   *
   * @returns Stats object
   */
  getStats(): MultimodalStats {
    let totalSize = 0;
    for (const f of this._fileMemories.values()) {
      totalSize += f.sizeBytes;
    }

    return {
      total_images: this._imageMemories.size,
      total_files: this._fileMemories.size,
      total_size_bytes: totalSize,
      image_descriptions: [...this._imageMemories.values()].map((m) => m.description),
    };
  }

  /**
   * Generate a unique ID from content using SHA-256.
   */
  private _generateId(content: string): string {
    return crypto.createHash("sha256").update(content, "utf-8").digest("hex").slice(0, 16);
  }
}

// Global singleton instance
let _multimodalStore: MultimodalMemoryStore | null = null;

/**
 * Get the global multimodal storage instance.
 *
 * @param basePath - Base path for storage
 * @returns MultimodalMemoryStore instance
 */
export function getMultimodalStore(basePath: string = "./workspace/multimodal"): MultimodalMemoryStore {
  if (_multimodalStore === null) {
    _multimodalStore = new MultimodalMemoryStore(basePath);
  }
  return _multimodalStore;
}

/**
 * Reset the global multimodal store instance (for testing).
 */
export function resetMultimodalStore(): void {
  _multimodalStore = null;
}
